### PR TITLE
Add fuzzy search

### PR DESCRIPTION
**Pull request by hendrasan**
@pveyes @tibudiyanto

Continuing from #14 (sorry, accidentally closed it due to my deleting the already pushed `main` branch since there were some errors and stuff when I tried to merge from latest, nevermind)

Anyway, I'm not sure about what Fuse options to use for the best experience. Searching "baru" now returns the Kalibaru, Marunda, and Semper Barat results (because of the fuzziness). It might be good because user's typos will still generate results, but it can also be a bit confusing for that "baru" case.

I only include the `nama_lokasi_vaksinasi`, wilayah, kecamatan, and kelurahan. I noticed there is `alamat_lokasi_vaksinasi` in the API, but since it was not shown in the frontend, I did not include it for the search.

### CODE REVIEW

File: src/pages/index.tsx [hunk 1]
```diff
@@ -156,15 +160,11 @@ export default function HomePage({ schedule }) {
...
-          <Select maxW={['auto', '2xs']} onChange={e => setSearchBy(e.target.value)} value={searchBy}>
-            <option value="kecamatan">Kecamatan</option>
-            <option value="kelurahan">Kelurahan</option>
-          </Select>
           <Input
             flexGrow={1}
             fontSize={[14, 16]}
             onChange={e => setSearchKeyword(e.target.value)}
-            placeholder="cari kecamatan/kelurahan"
+            placeholder="cari nama lokasi vaksinasi/wilayah/kecamatan/kelurahan"
```
Comment (pveyes) [inline src/pages/index.tsx]: I think wilayah add confusion, we can just use lokasi / kecamatan / kelurahan

Also reorder the list so kecamatan & kelurahan go first, as it's more common

```suggestion
            placeholder="cari kecamatan / kelurahan / nama lokasi vaksinasi"
```

File: src/pages/index.tsx [hunk 2]
```diff
@@ -48,15 +49,20 @@ export default function HomePage({ schedule }) {
       return schedule;
     }
 
+    const fuse = new Fuse(schedule, {
+      keys: ['nama_lokasi_vaksinasi', 'wilayah', 'kecamatan', 'kelurahan'],
```
Comment (pveyes) [inline src/pages/index.tsx]: @tibudiyanto  should we add weighted keys? `kecamatan` and `kelurahan` is more commonly searched
```suggestion
      keys: [
       {
         name: 'kecamatan',
         weight: 2,
       },
       {
         name: 'kelurahan',
         weight: 2,
       }
       // not sure if this works but default fuse default weight is 1 
      'nama_lokasi_vaksinasi',
      'wilayah'
      ],
```

File: src/pages/index.tsx [hunk 3]
```diff
@@ -48,15 +49,20 @@ export default function HomePage({ schedule }) {
...
+      threshold: 0.4
+    });
+
+    const results = fuse.search(searchKeyword);
```
Comment (pveyes) [inline src/pages/index.tsx]: how's the perf? maybe we can add memo if needed here

```suggestion
    const results = useMemo(() => fuse.search(searchKeyword), [searchKeyword]);
```

**Review by pveyes — APPROVED** [review]
Overall LGTM, I'm adding @tibudiyanto & @resir014 as reviewer

File: src/pages/index.tsx (hunk 2 above)
Comment (tibudiyanto) [reply src/pages/index.tsx]: I think this is great

**Comment by tibudiyanto** [comment]
sorry you have conflicts :(

**Comment by pveyes** [comment]
also, can you rebase after #56 and add the same fuzzy search logic to map page?

Maybe create both custom component `<SearchBox />` and custom hook `useFuzzySearch`

Just to illustrate
```tsx
const filtered = useFuzzySearch(schedule, searchQuery)

<SearchBox query={searchQuery} onChange={query => setSearchQuery(query)} />
```

File: src/pages/index.tsx (hunk 1 above)
Comment (mazipan) [inline src/pages/index.tsx]: Small fix, use `Cari` not `cari`.
No need a space on `/` char

**Comment by hendrasan** [comment]
Okay, updated. Should hopefully work and not break stuff.

File: src/pages/index.tsx [hunk 4]
```diff
@@ -129,12 +129,10 @@ export default function HomePage({ schedule }: Props) {
...
     }
 
-    return schedule.filter(s => {
-      return s[searchBy].toLowerCase().includes(searchKeyword.toLowerCase());
-    });
+    return filteredSchedules;
     // Resorting to a typecast here for a quick workaround.
     // TODO: Type this function better
-  }, [schedule, searchBy, searchKeyword, userLocation]) as VaccinationDataWithDistance[];
+  }, [schedule, searchKeyword, filtered, userLocation]) as VaccinationDataWithDistance[];
```
Comment (pveyes) [inline src/pages/index.tsx:135]: just realized that this memo is not ideal, too many computed values (`filtered` is computed value from `schedule` and `searchKeyword`). let's fix this in the separate PR

File: src/pages/map.tsx [hunk 5]
```diff
@@ -74,23 +75,21 @@ interface Props {
...
     }
 
-    const result = schedule.filter(props => {
-      const fieldValue = props[searchBy].toLowerCase();
+    const filteredSchedules = searchKeyword ? filtered.map(s => s.item) : schedule;
 
-      return (
-        fieldValue.includes(searchKeyword.toLowerCase()) && props.detail_lokasi != null && props.detail_lokasi.length
-      );
+    const result = filteredSchedules.filter(props => {
```
Comment (pveyes) [inline src/pages/map.tsx:112]: same with this logic, this can be extracted and reused between index/map pages. TODO for #57

**Review by pveyes — APPROVED** [review]
Tested locally. LGTM homepage & impl. Nice work!

cc @sonnylazuardi for map

**Comment by sonnylazuardi** [comment]
image

please rebase, i think there should be current location button on map

**Comment by pveyes** [comment]
weird iOS 15 flex, but okay 🤪

File: src/pages/index.tsx (hunk 4 above)
Comment (hendrasan) [reply src/pages/index.tsx:135]: yeah, i'm not that familiar with React and Typescript, so maybe you or the others can tackle optimizing this later.

**Comment by hendrasan** [comment]
Updated

**Review by sonnylazuardi — APPROVED** [review]
lgtm

**Review by pveyes — APPROVED** [review]
LGTM, thanks for making this. Let's 🚢 it

**Pull request merged**